Add documentation for smart contract upgradeability (#638)
Concordium/concordium.github.io: source/mainnet/smart-contracts/general/contract-instances.rst

handle, and the state and balance of the instance remain unchanged.
The account which sent the initiating transaction pays for the execution of the
entire receive function, including the cost of failed operations.

.. _contract-instance-upgradeability:

Upgradeability
==============

Smart contract instances can choose to upgrade its module to a new smart contract
module using the **upgrade** host function.
The host function takes a reference to a deployed smart contract module to use for
the upgraded instance and can only be called from a receive function.

> The function returns back whether the upgrade is successful or failed, allowing the instance

Reviewer: ```suggestion
The host function returns whether the upgrade succeeded, allowing the instance
```

> to decide the next step. If the upgrade is succeful any new invocations of the

Reviewer: ```suggestion
to decide the next step. If the upgrade is successful, any new invocations of the
```

upgraded instance uses the smart contract code in the new module.

.. warning::

>    Make sure to proper authorized any endpoint triggering a smart contract upgrade as
>    it allows for changing the behavior of the smart contract.

Reviewer: Perhaps?
```suggestion
   Make sure to restrict access to any endpoint triggering a smart contract upgrade as
   it allows for changing the behavior of the smart contract.
```

>    it allows for changing the behavior of the smart contract.

Author: I have tried to rephrase this

.. graphviz::
   :align: center
   :caption: Example of a smart contract instance 'Car' upgrading to a new module.

   digraph G {
       rankdir="BT"

       subgraph cluster_0 {
           label = "My module v1";
           labelloc=b;
           node [fillcolor=white, shape=note]
           escrow;
       }

       subgraph cluster_2 {
           label = "My module v2";
           labelloc=b;
           node [fillcolor=white, shape=note]
           escrow2;
       }

       subgraph cluster_1 {
           label = "Instances";
           style=dotted;
           node [shape=box, style=rounded]
           House;
           Car;
       }

       escrow[label="Escrow"]
       escrow2[label="Escrow"]

       House:n -> escrow;

>        Car:n -> escrow [style=dashed];

Reviewer: I think `dashed` is a bit hard to distinguish from the solid line. Consider using a different style.

Author: Unfortunately, there are not that many options, but I have tried some other things to make it stand out a bit more.
Let me know what you think

Reviewer: Your changes made it much better. Great!

       Car:n -> escrow2;
   }

Failing to upgrade
------------------

A smart contract instance can fail to upgrade for one of the following reasons:

- The new module does not exist.
- The new module does not contain a smart contract with a name matching the instance being upgraded.
- The new module is a smart contract module version 0.

Immutability
------------

It is still possible to implement an immutable smart contract, by simply `not` implementing any logic

> triggering the *upgrade* host function.

Reviewer 2: Can we make this even more prominent?

Author: I have tried to improve this

.. graphviz::
   :align: center
   :caption: Example of a smart contract instance 'Car' upgrading to a new module.

   digraph G {
       rankdir="BT"

       subgraph cluster_0 {
           label = "My module v1";
           labelloc=b;
           node [fillcolor=white, shape=note]
           escrow;
       }

       subgraph cluster_2 {
           label = "My module v2";
           labelloc=b;
           node [fillcolor=white, shape=note]
           escrow2;
       }

       subgraph cluster_1 {
           label = "Instances";
           style=dotted;
           node [shape=box, style=rounded]
           House;
           Car;
       }

       escrow[label="Escrow"]
       escrow2[label="Escrow"]

       House:n -> escrow;
       Car:n -> escrow [style=dashed, arrowhead=onormal];
       Car:n -> escrow2 [style=bold];
   }

Failing to upgrade
------------------

A smart contract instance can fail to upgrade for one of the following reasons:

- The new module does not exist.
- The new module does not contain a smart contract with a name matching the instance being upgraded.
- The new module is a smart contract module version 0.

Immutability
------------

> Only the smart contract itself can trigger an upgrade of its module, meaning that smart contracts
> are immutable, when it does not contain any code for triggering an upgrade.

Reviewer 2: Can we be a bit more precise here and also add a warning that just because they cannot be upgraded does not mean that the behaviour of a contract cannot change due to state changes, for example.
```suggestion
Only the smart contract itself can trigger an upgrade of its module, meaning that smart contracts
are immutable, when they do not contain any code for triggering an upgrade.
```

Author: I have added a warning
